File: app/src/main/java/com/qimian233/ztool/hook/modules/systemui/CustomDateFormatter.java
package com.qimian233.ztool.hook.modules.systemui;

import android.util.Log;

import com.nlf.calendar.Lunar;
import com.nlf.calendar.Solar;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * 自定义日期格式化工具类
 * 支持农历、节气、时辰等特殊格式
 */
public class CustomDateFormatter {

    private static final String TAG = "CustomDateFormatter";

    // 时辰对照表
    private static final String[] CHINESE_HOURS = {
            "子时", "丑时", "寅时", "卯时", "辰时", "巳时",
            "午时", "未时", "申时", "酉时", "戌时", "亥时"
    };

    // 地支对照表
    private static final String[] EARTHLY_BRANCHES = {
            "子", "丑", "寅", "卯", "辰", "巳",
            "午", "未", "申", "酉", "戌", "亥"
    };

    // 自定义格式模式映射
    private static final Map<String, String> CUSTOM_PATTERNS = new HashMap<>();

    static {
        // 初始化自定义模式映射
        CUSTOM_PATTERNS.put("N", "lunar");           // 农历日期
        CUSTOM_PATTERNS.put("J", "solarTerm");       // 节气
        CUSTOM_PATTERNS.put("T", "chineseHour");     // 时辰
        CUSTOM_PATTERNS.put("C", "constellation");   // 星座
        CUSTOM_PATTERNS.put("A", "animal");          // 生肖
        CUSTOM_PATTERNS.put("W", "week");            // 星期
    }

    /**
     * 格式化日期，支持自定义农历、节气等格式
     *
     * @param pattern 格式模式，支持以下自定义占位符：
     *                N - 农历日期（如：腊月廿三）
     *                J - 节气（如：立春）
     *                T - 时辰（如：子时）
     *                C - 星座（如：水瓶座）
     *                A - 生肖（如：龙）
     *                W - 星期（如：星期一）
     *                同时支持标准的SimpleDateFormat格式
     *
     * @param date 要格式化的日期
     * @return 格式化后的字符串
     */
    public static String format(String pattern, Date date) {
        if (pattern == null || date == null) {
            return "";
        }

        try {
            // 处理自定义格式
            String result = processCustomPatterns(pattern, date);

            // 处理标准的SimpleDateFormat格式
            result = processStandardPatterns(result, date);

            return result;

        } catch (Exception e) {
            Log.e(TAG, "Error formatting date: " + e.getMessage(), e);
            // 出错时返回默认格式
            return new SimpleDateFormat("HH:mm", Locale.getDefault()).format(date);
        }
    }

    /**
     * 处理自定义模式（农历、节气、时辰等）
     */
    private static String processCustomPatterns(String pattern, Date date) {
        String result = pattern;
        Lunar lunar = Lunar.fromDate(date);

        // 逐个处理自定义占位符
        for (Map.Entry<String, String> entry : CUSTOM_PATTERNS.entrySet()) {
            String placeholder = entry.getKey();
            String type = entry.getValue();

            if (result.contains(placeholder)) {
                String replacement = getCustomReplacement(type, lunar, date);
                result = result.replace(placeholder, replacement);
            }
        }

        return result;
    }

    /**
     * 处理标准SimpleDateFormat格式
     */
    private static String processStandardPatterns(String pattern, Date date) {
        // 如果已经没有自定义占位符，直接格式化
        if (!containsCustomPatterns(pattern)) {
            return new SimpleDateFormat(pattern, Locale.getDefault()).format(date);
        }

        // 使用正则表达式匹配并替换标准格式部分
        StringBuilder result = new StringBuilder();
        Pattern stdPattern = Pattern.compile("([^a-zA-Z]|^)([yMdHhmsSEDFwWkKzZ]+)([^a-zA-Z]|$)");
        Matcher matcher = stdPattern.matcher(pattern);

        int lastEnd = 0;
        while (matcher.find()) {
            // 添加非格式部分
            result.append(pattern, lastEnd, matcher.start(2));

            // 格式化标准部分
            String stdFormat = matcher.group(2);
            String formatted = new SimpleDateFormat(stdFormat, Locale.getDefault()).format(date);
            result.append(formatted);

            lastEnd = matcher.end(2);
        }

        // 添加剩余部分
        result.append(pattern.substring(lastEnd));

        return result.toString();
    }

    /**
     * 获取自定义占位符的替换内容
     */
    private static String getCustomReplacement(String type, Lunar lunar, Date date) {
        switch (type) {
            case "lunar":
                return getLunarDate(lunar);
            case "solarTerm":
                return getSolarTerm(lunar);
            case "chineseHour":
                return getChineseHour(date);
            case "constellation":
                return getConstellation(lunar);
            case "animal":
                return lunar.getYearShengXiao();
            case "week":
                return getChineseWeek(date);
            default:
                return "";
        }
    }

    /**
     * 获取农历日期
     */
    private static String getLunarDate(Lunar lunar) {
        try {
            // 格式：腊月廿三
            return lunar.getMonthInChinese() + "月" + lunar.getDayInChinese();
        } catch (Exception e) {
            Log.e(TAG, "Error getting lunar date", e);
            return "";
        }
    }

    /**
     * 获取节气
     */
    private static String getSolarTerm(Lunar lunar) {
        try {
            String solarTerm = lunar.getJieQi();
            return solarTerm != null ? solarTerm : "";
        } catch (Exception e) {
            Log.e(TAG, "Error getting solar term", e);
            return "";
        }
    }

    /**
     * 获取时辰
     */
    private static String getChineseHour(Date date) {
        try {
            SimpleDateFormat hourFormat = new SimpleDateFormat("HH", Locale.getDefault());
            int hour = Integer.parseInt(hourFormat.format(date));

            // 计算时辰（每2小时一个时辰）
            int hourIndex = (hour + 1) / 2 % 12;
            return CHINESE_HOURS[hourIndex];

        } catch (Exception e) {
            Log.e(TAG, "Error getting Chinese hour", e);
            return "";
        }
    }

    /**
     * 获取星座
     */
    private static String getConstellation(Lunar lunar) {
        try {
            Solar solar = lunar.getSolar();
            int month = solar.getMonth();
            int day = solar.getDay();

            return getConstellationBySolarDate(month, day);

        } catch (Exception e) {
            Log.e(TAG, "Error getting constellation", e);
            return "";
        }
    }

    /**
     * 根据公历日期获取星座
     */
    private static String getConstellationBySolarDate(int month, int day) {
        if ((month == 1 && day >= 20) || (month == 2 && day <= 18)) return "水瓶座";
        if ((month == 2 && day >= 19) || (month == 3 && day <= 20)) return "双鱼座";
        if ((month == 3 && day >= 21) || (month == 4 && day <= 19)) return "白羊座";
        if ((month == 4 && day >= 20) || (month == 5 && day <= 20)) return "金牛座";
        if ((month == 5 && day >= 21) || (month == 6 && day <= 21)) return "双子座";
        if ((month == 6 && day >= 22) || (month == 7 && day <= 22)) return "巨蟹座";
        if ((month == 7 && day >= 23) || (month == 8 && day <= 22)) return "狮子座";
        if ((month == 8 && day >= 23) || (month == 9 && day <= 22)) return "处女座";
        if ((month == 9 && day >= 23) || (month == 10 && day <= 23)) return "天秤座";
        if ((month == 10 && day >= 24) || (month == 11 && day <= 22)) return "天蝎座";
        if ((month == 11 && day >= 23) || (month == 12 && day <= 21)) return "射手座";
        if ((month == 12 && day >= 22) || (month == 1 && day <= 19)) return "摩羯座";
        return "";
    }

    /**
     * 获取中文星期
     */
    private static String getChineseWeek(Date date) {
        try {
            SimpleDateFormat weekFormat = new SimpleDateFormat("E", Locale.CHINA);
            String week = weekFormat.format(date);
            return week.replace("星期", "周"); // 统一格式为"周一"
        } catch (Exception e) {
            Log.e(TAG, "Error getting Chinese week", e);
            return "";
        }
    }

    /**
     * 检查是否包含自定义模式
     */
    private static boolean containsCustomPatterns(String pattern) {
        for (String placeholder : CUSTOM_PATTERNS.keySet()) {
            if (pattern.contains(placeholder)) {
                return true;
            }
        }
        return false;
    }

    /**
     * 获取所有支持的自定义格式说明
     */
    public static String getSupportedFormats() {
        return "支持的自定义格式:\n" +
                "N - 农历日期 (如: 腊月廿三)\n" +
                "J - 节气 (如: 立春)\n" +
                "T - 时辰 (如: 子时)\n" +
                "C - 星座 (如: 水瓶座)\n" +
                "A - 生肖 (如: 龙)\n" +
                "W - 星期 (如: 周一)\n" +
                "同时支持标准日期格式: yyyy, MM, dd, HH, mm, ss等";
    }
}

File: app/src/main/java/com/qimian233/ztool/hook/modules/systemui/misc/CustomDateFormatter.kt
package com.qimian233.ztool.hook.modules.systemui.misc

import android.util.Log
import com.nlf.calendar.Lunar
import com.nlf.calendar.Solar
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.regex.Pattern

/**
 * Custom date formatter utility.
 * Supports special formats such as lunar calendar, solar terms, Chinese hours, and time periods.
 */
object CustomDateFormatter {

    private const val TAG = "CustomDateFormatter"

    // Chinese traditional hour (shichen) lookup table
    private val CHINESE_HOURS = arrayOf(
        "子时", "丑时", "寅时", "卯时", "辰时", "巳时",
        "午时", "未时", "申时", "酉时", "戌时", "亥时"
    )

    // Custom format pattern mapping
    private val CUSTOM_PATTERNS = mapOf(
        "N" to "lunar",           // lunar date
        "J" to "solarTerm",       // solar term
        "T" to "chineseHour",     // Chinese hour (shichen)
        "C" to "constellation",   // constellation
        "A" to "animal",          // zodiac animal
        "W" to "week",            // week
        "a" to "timePeriod"       // time period
    )

    /**
     * Format the date, supporting custom lunar calendar, solar term, etc. formats.
     *
     * @param pattern format pattern; supports the following custom placeholders:
     *                N - lunar date (e.g.: 腊月廿三)
     *                J - solar term (e.g.: 立春)
     *                T - Chinese hour (e.g.: 子时)
     *                C - constellation (e.g.: 水瓶座)
     *                A - zodiac animal (e.g.: 龙)
     *                W - week (e.g.: 星期一)
     *                a - time period (e.g.: 凌晨、上午、中午、下午、晚上)
     *                Standard SimpleDateFormat patterns are also supported.
     *
     * @param date the date to format
     * @return the formatted string
     */
    fun format(pattern: String?, date: Date?): String {
        if (pattern == null || date == null) {
            return ""
        }

        return try {
            // Process custom patterns
            var result = processCustomPatterns(pattern, date)

            // Process standard SimpleDateFormat patterns
            result = processStandardPatterns(result, date)

            result
        } catch (e: Exception) {
            Log.e(TAG, "Error formatting date: " + e.message, e)
            // Return the default format on error
            SimpleDateFormat("HH:mm", Locale.getDefault()).format(date)
        }
    }

    /**
     * Process custom patterns (lunar calendar, solar terms, Chinese hours, etc.)
     */
    private fun processCustomPatterns(pattern: String, date: Date): String {
        var result = pattern
        val lunar = Lunar.fromDate(date)

        // Process custom placeholders one by one
        for ((placeholder, type) in CUSTOM_PATTERNS) {
            if (result.contains(placeholder)) {
                val replacement = getCustomReplacement(type, lunar, date)
                result = result.replace(placeholder, replacement)
            }
        }

        return result
    }

    /**
     * Process standard SimpleDateFormat patterns
     */
    private fun processStandardPatterns(pattern: String, date: Date): String {
        // If no custom placeholders remain, format directly
        if (!containsCustomPatterns(pattern)) {
            return SimpleDateFormat(pattern, Locale.getDefault()).format(date)
        }

        // Match and replace standard format segments using regex
        val result = StringBuilder()
        val stdPattern = Pattern.compile("([^a-zA-Z]|^)([yMdHhmsSEDFwWkKzZ]+)([^a-zA-Z]|$)")
        val matcher = stdPattern.matcher(pattern)

        var lastEnd = 0
        while (matcher.find()) {
            // Append non-format segments
            result.append(pattern, lastEnd, matcher.start(2))

            // Format the standard segment
            val stdFormat = matcher.group(2)
            val formatted = SimpleDateFormat(stdFormat, Locale.getDefault()).format(date)
            result.append(formatted)

            lastEnd = matcher.end(2)
        }

        // Append the remaining part
        result.append(pattern.substring(lastEnd))

        return result.toString()
    }

    /**
     * Get the replacement content for a custom placeholder
     */
    private fun getCustomReplacement(type: String, lunar: Lunar, date: Date): String {
        return when (type) {
            "lunar" -> getLunarDate(lunar)
            "solarTerm" -> getSolarTerm(lunar)
            "chineseHour" -> getChineseHour(date)
            "constellation" -> getConstellation(lunar)
            "animal" -> lunar.yearShengXiao
            "week" -> getChineseWeek(date)
            "timePeriod" -> getTimePeriod(date)
            else -> ""
        }
    }

    /**
     * Get the lunar date
     */
    private fun getLunarDate(lunar: Lunar): String {
        return try {
            // Format: 腊月廿三
            lunar.monthInChinese + "月" + lunar.dayInChinese
        } catch (e: Exception) {
            Log.e(TAG, "Error getting lunar date", e)
            ""
        }
    }

    /**
     * Get the solar term
     */
    private fun getSolarTerm(lunar: Lunar): String {
        return try {
            val solarTerm = lunar.jieQi
            solarTerm ?: ""
        } catch (e: Exception) {
            Log.e(TAG, "Error getting solar term", e)
            ""
        }
    }

    /**
     * Get the Chinese hour (shichen)
     */
    private fun getChineseHour(date: Date): String {
        return try {
            val hourFormat = SimpleDateFormat("HH", Locale.getDefault())
            val hour = hourFormat.format(date).toInt()

            // Compute the shichen (one shichen per 2 hours)
            val hourIndex = (hour + 1) / 2 % 12
            CHINESE_HOURS[hourIndex]
        } catch (e: Exception) {
            Log.e(TAG, "Error getting Chinese hour", e)
            ""
        }
    }

    /**
     * Get the constellation
     */
    private fun getConstellation(lunar: Lunar): String {
        return try {
            val solar: Solar = lunar.solar
            val month = solar.month
            val day = solar.day

            getConstellationBySolarDate(month, day)
        } catch (e: Exception) {
            Log.e(TAG, "Error getting constellation", e)
            ""
        }
    }

    /**
     * Get the constellation from a Gregorian date
     */
    private fun getConstellationBySolarDate(month: Int, day: Int): String {
        if (month == 1 && day >= 20 || month == 2 && day <= 18) return "水瓶座"
        if (month == 2 || month == 3 && day <= 20) return "双鱼座"
        if (month == 3 || month == 4 && day <= 19) return "白羊座"
        if (month == 4 || month == 5 && day <= 20) return "金牛座"
        if (month == 5 || month == 6 && day <= 21) return "双子座"
        if (month == 6 || month == 7 && day <= 22) return "巨蟹座"
        if (month == 7 || month == 8 && day <= 22) return "狮子座"
        if (month == 8 || month == 9 && day <= 22) return "处女座"
        if (month == 9 || month == 10 && day <= 23) return "天秤座"
        if (month == 10 || month == 11 && day <= 22) return "天蝎座"
        if (month == 11 || month == 12 && day <= 21) return "射手座"
        if (month == 12 || month == 1) return "摩羯座"
        return ""
    }

    /**
     * Get the Chinese week name
     */
    private fun getChineseWeek(date: Date): String {
        return try {
            val weekFormat = SimpleDateFormat("E", Locale.CHINA)
            val week = weekFormat.format(date)
            week.replace("星期", "周") // Normalize to "周一" style
        } catch (e: Exception) {
            Log.e(TAG, "Error getting Chinese week", e)
            ""
        }
    }

    /**
     * Get the current time period (early morning, morning, afternoon, etc.)
     */
    private fun getTimePeriod(date: Date): String {
        return try {
            val hourFormat = SimpleDateFormat("H", Locale.getDefault())
            val hour = hourFormat.format(date).toInt()

            when (hour) {
                in 0..<6 -> "凌晨"
                in 6..<9 -> "早上"
                in 9..<12 -> "上午"
                in 12..<14 -> "中午"
                in 14..<18 -> "下午"
                in 18..<24 -> "晚上"
                else -> ""
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error getting time period", e)
            ""
        }
    }

    /**
     * Check whether the pattern contains custom patterns
     */
    private fun containsCustomPatterns(pattern: String): Boolean {
        for (placeholder in CUSTOM_PATTERNS.keys) {
            if (pattern.contains(placeholder)) {
                return true
            }
        }
        return false
    }
}
